smdebug/tensorflow/keras.py

```diff
@@ -620,3 +621,75 @@ def new_add_weight(opt, *args, **kwargs):
         # Optimizer is being saved to support additional features in the future.
         self.optimizer = optimizer
         return optimizer
+
+    def forward_pre_hook(self, tape):
+        self.tape = tape
+        self.set_mode(ModeKeys.TRAIN)
+        if self._is_not_supported():
+            return
+        self.worker = self._get_worker_name()
+        if self.writer is not None or len(self.writer_map):
+            self._close_writers()
+        if not self.prepared_collections:
+            # at this point we need all collections to be ready
+            # this may not be the case at creation of hook
+            # as user's code after hook might add collections
+            self._prepare_collections()
+            self.prepared_collections = True
+
+        self._increment_step()
+
+        if self._get_collections_to_save_for_step():
+            self._initialize_writers()
+
+        if self._exported_collections is False:
+            # in keras, these collections change when mode changes
+            # but rest of the project isn't yet capable of handling this
+            # this means that collections like outputs, or other collections with intermediate tensors
+            # will only have tensor names from first mode
+
+            # this means sometimes collections will be exported after 1 step
+            self.export_collections()
+            self._exported_collections = True
```

Comment:
This seems like a pretty important gotcha when users run both training and validation each epoch. Let's brainstorm more how to improve the customer experience in this case, don't want users to be confused.